File: src/posture_game/posture_game/checker_node.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
from std_msgs.msg import Bool, Float32MultiArray, Float32
from posture_game.postures import PostureDetector
import time
import cv2
import mediapipe as mp
import threading
#from rclpy.qos import qos_profile_sensor_data
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup
from threading import Lock



class CheckerNode(Node):

    def __init__(self):
        super().__init__('checker_node')

        self.bridge = CvBridge()
        self.latest_frame = None
        self.detector = PostureDetector()
        self.last_frame_timestamp = None
        self.callback_group = ReentrantCallbackGroup()
        self.lock_presence = Lock()
        self.lock_validation = Lock()


        mp_holistic = mp.solutions.holistic

        self.holistic_presence = mp_holistic.Holistic(
            static_image_mode=False,
            model_complexity=1,
            smooth_landmarks=True,
            enable_segmentation=False,
            refine_face_landmarks=False,
            min_detection_confidence=0.5,
            min_tracking_confidence=0.5
            )

        self.holistic_validate = mp_holistic.Holistic(
            static_image_mode=False,
            model_complexity=1,
            smooth_landmarks=True,
            enable_segmentation=False,
            refine_face_landmarks=False,
            min_detection_confidence=0.5,
            min_tracking_confidence=0.5
            )



        

        # Subscripciones
        self.create_subscription(Float32MultiArray,'/current_pose_data',self.validate_pose_callback,10,callback_group=self.callback_group)
        self.create_subscription(Image,'/image_raw',self.frame_callback,10,callback_group=self.callback_group)




        # Publicaciones
        self.result_publisher = self.create_publisher(Bool, '/pose_result', 10)
        self.duration_pub = self.create_publisher(Float32, '/pose_duration', 10)
        self.presence_pub = self.create_publisher(Bool, '/player_present', 10)


        # Variables de validación
        self.validation_thread = None
        self.stop_event = threading.Event()

        #Timer
        self.presence_timer = self.create_timer(0.5, self.check_presence)




    def check_presence(self):
        if self.latest_frame is None:
            return

        try:
            frame_cv2 = self.bridge.imgmsg_to_cv2(self.latest_frame)
            bgr = cv2.cvtColor(frame_cv2, cv2.COLOR_YUV2BGR_YUY2)
            rgb = cv2.cvtColor(bgr, cv2.COLOR_BGR2RGB)

            with self.lock_presence:
                results = self.holistic_presence.process(rgb)

            msg_out = Bool()
            msg_out.data = results.pose_landmarks is not None
            self.presence_pub.publish(msg_out)

        except Exception as e:
            self.get_logger().warn(f"[Presencia] Error procesando frame: {e}")

    def frame_callback(self, msg):
        self.latest_frame = msg
        #self.last_frame_timestamp = msg.header.stamp.sec + msg.header.stamp.nanosec * 1e-9

    def validate_pose_callback(self, msg):
        if self.latest_frame is None:
            self.get_logger().warn("⚠️ No hay imagen recibida todavía")
            return

        try:
            pose_id = int(msg.data[0])
            timeout = float(msg.data[1])
        except (IndexError, ValueError) as e:
            self.get_logger().error(f"❌ Datos inválidos: {msg.data} - Error: {str(e)}")
            return

        self.get_logger().info(f"🧠 Iniciando validación de postura {pose_id} durante {timeout} segundos")

        # Detener cualquier validación anterior
        if self.validation_thread and self.validation_thread.is_alive():
            self.stop_event.set()
            self.validation_thread.join()

        # Iniciar nueva validación
        self.stop_event.clear()
        self.validation_thread = threading.Thread(
            target=self.run_validation_loop,
            args=(pose_id, timeout),
            daemon=True
        )
        self.validation_thread.start()

    def run_validation_loop(self, pose_id, timeout):


        start_time = time.time()
        
        # 🛑 Esperar que llegue un frame nuevo
        #initial_ts = self.last_frame_timestamp
        #wait_limit = 2.0  # segundos máximos de espera
        #while self.last_frame_timestamp == initial_ts:
            #if time.time() - start_time > wait_limit:
               # self.get_logger().warn("⏱️ No llegó un nuevo frame para iniciar validación")
                #self.publish_result(False, 0.0)
               # return
          #  time.sleep(0.05)

        while not self.stop_event.is_set():
            elapsed = time.time() - start_time
            if elapsed > timeout:
                self.get_logger().info("⏱️ Tiempo de validación finalizado sin éxito")
                self.publish_result(False, elapsed)
                return

            frame = self.latest_frame
            if frame is None:
                continue

            try:
                frame_cv2 = self.bridge.imgmsg_to_cv2(frame)
                if frame_cv2 is None or frame_cv2.size == 0:
                    self.get_logger().warn("[Checker] Frame vacío o inválido")
                    continue


                bgr = cv2.cvtColor(frame_cv2, cv2.COLOR_YUV2BGR_YUY2)
                rgb = cv2.cvtColor(bgr, cv2.COLOR_BGR2RGB)
            except Exception as e:
                self.get_logger().error(f"[ERROR] Conversión de imagen: {e}")
                continue
            
            with self.lock_validation:
                results = self.holistic_validate.process(rgb)

            
            pose_ok = False
            hand_ok = False

            if results.pose_landmarks:
                lm = results.pose_landmarks.landmark

                if pose_id == 0:
                    pose_ok = self.detector.is_right_arm_up(lm)
                elif pose_id == 1:
                    pose_ok = self.detector.is_left_arm_up(lm)
                elif pose_id == 2:
                    pose_ok = self.detector.are_arms_in_x(lm)
                elif pose_id == 6:
                    pose_ok = self.detector.is_right_arm_horizontal(lm)
                elif pose_id == 7:
                    pose_ok = self.detector.is_left_arm_horizontal(lm)
                elif pose_id == 8:
                    pose_ok = self.detector.are_both_arms_horizontal(lm)
                elif pose_id == 9:
                    pose_ok = self.detector.is_wrist_touching_nose(lm, "right") or self.detector.is_wrist_touching_nose(lm, "left")

            # Puño derecho solamente
            if pose_id == 3 and results.right_hand_landmarks:
                hand_ok = self.detector.is_right_fist_exclusive(
                    results.right_hand_landmarks,
                    results.left_hand_landmarks if results.left_hand_landmarks else None
                )

            # Puño izquierdo solamente
            elif pose_id == 4 and results.left_hand_landmarks:
                hand_ok = self.detector.is_left_fist_exclusive(
                    results.left_hand_landmarks,
                    results.right_hand_landmarks if results.right_hand_landmarks else None
                )

            # Ambos puños
            elif pose_id == 5 and results.right_hand_landmarks and results.left_hand_landmarks:
                hand_ok = self.detector.are_both_fists_closed(
                    results.right_hand_landmarks,
                    results.left_hand_landmarks
                )

            #Verificar validacion
            if pose_ok or hand_ok:
                self.get_logger().info("✅ Postura detectada correctamente")
                duration = time.time() - start_time
                self.publish_result(True, duration)
                return

            time.sleep(0.1)  # evita uso excesivo de CPU

    def publish_result(self, result: bool, duration: float):
        result_msg = Bool()
        result_msg.data = result
        self.result_publisher.publish(result_msg)

        duration_msg = Float32()
        duration_msg.data = duration
        self.duration_pub.publish(duration_msg)

        estado = "✅ Correcto" if result else "❌ Incorrecto"
        self.get_logger().info(f"📤 Resultado publicado: {estado} - Tiempo: {duration:.2f}s")
        self.get_logger().info(f"-------------------------------------------------")

def main(args=None):
    rclpy.init(args=args)
    node = CheckerNode()
    executor = MultiThreadedExecutor()
    executor.add_node(node)
    try:
        executor.spin()
    except KeyboardInterrupt:
        pass
    finally:
        node.holistic_presence.close()
        node.holistic_validate.close()
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: posture_game/src/checker_node.cpp
#include "posture_game/checker_node.hpp"

using namespace std::chrono_literals;
using std::placeholders::_1;

namespace posture_game
{

CheckerNode::CheckerNode() : Node("checker_node"), detector_()
{
    result_pub_ = this->create_publisher<std_msgs::msg::Bool>("/pose_result", 10);
    duration_pub_ = this->create_publisher<std_msgs::msg::Float32>("/pose_duration", 10);
    presence_pub_ = this->create_publisher<std_msgs::msg::Bool>("/player_present", 10);

    pose_data_sub_ = this->create_subscription<std_msgs::msg::Float32MultiArray>(
        "/current_pose_data", 10,
        std::bind(&CheckerNode::validate_pose_callback, this, _1));

    landmarks_sub_ = this->create_subscription<posture_game_interfaces::msg::HolisticLandmarks>(
        "/holistic_landmarks", 10,
        std::bind(&CheckerNode::landmarks_callback, this, _1));

    presence_timer_ = this->create_wall_timer(
        500ms, std::bind(&CheckerNode::check_presence, this));

    RCLCPP_INFO(this->get_logger(), "✅ Nodo CheckerNode en ejecución.");
}

void CheckerNode::landmarks_callback(const posture_game_interfaces::msg::HolisticLandmarks::SharedPtr msg)
{
    std::lock_guard<std::mutex> lock(landmarks_mutex_);
    pose_landmarks_ = parse_landmarks(msg->pose_landmarks);
    left_hand_landmarks_ = parse_landmarks(msg->left_hand_landmarks);
    right_hand_landmarks_ = parse_landmarks(msg->right_hand_landmarks);
}

void CheckerNode::check_presence()
{
    std::lock_guard<std::mutex> lock(landmarks_mutex_);
    bool present = !pose_landmarks_.empty();
    presence_pub_->publish(std_msgs::msg::Bool().set__data(present));
}

void CheckerNode::validate_pose_callback(const std_msgs::msg::Float32MultiArray::SharedPtr msg)
{
    if (msg->data.size() < 2)
    {
        RCLCPP_ERROR(this->get_logger(), "❌ Mensaje inválido recibido.");
        return;
    }

    int pose_id = static_cast<int>(msg->data[0]);
    float timeout = msg->data[1];

    RCLCPP_INFO(this->get_logger(), "🧠 Validando postura %d por %.1f s", pose_id, timeout);

    stop_flag_ = true;
    if (validation_thread_.joinable())
        validation_thread_.join();

    stop_flag_ = false;
    validation_thread_ = std::thread(&CheckerNode::run_validation_loop, this, pose_id, timeout);
}

void CheckerNode::run_validation_loop(int pose_id, float timeout)
{
    auto start = std::chrono::steady_clock::now();

    while (!stop_flag_)
    {
        auto now = std::chrono::steady_clock::now();
        float elapsed = std::chrono::duration<float>(now - start).count();

        if (elapsed > timeout)
        {
            RCLCPP_INFO(this->get_logger(), "⏱️ Tiempo agotado sin éxito.");
            publish_result(false, elapsed);
            return;
        }

        std::vector<MockLandmark> lm, rh, lh;
        {
            std::lock_guard<std::mutex> lock(landmarks_mutex_);
            lm = pose_landmarks_;
            rh = right_hand_landmarks_;
            lh = left_hand_landmarks_;
        }

        bool pose_ok = false, hand_ok = false;

        if (!lm.empty())
        {
            if (pose_id == 0) pose_ok = detector_.is_right_arm_up(lm);
            else if (pose_id == 1) pose_ok = detector_.is_left_arm_up(lm);
            else if (pose_id == 2) pose_ok = detector_.are_arms_in_x(lm);
            else if (pose_id == 6) pose_ok = detector_.is_right_arm_horizontal(lm);
            else if (pose_id == 7) pose_ok = detector_.is_left_arm_horizontal(lm);
            else if (pose_id == 8) pose_ok = detector_.are_both_arms_horizontal(lm);
            else if (pose_id == 9) pose_ok = detector_.is_wrist_touching_nose(lm, "right") ||
                                            detector_.is_wrist_touching_nose(lm, "left");
        }

        if (pose_id == 3 && !rh.empty())
            hand_ok = detector_.is_right_fist_exclusive(rh, &lh);
        else if (pose_id == 4 && !lh.empty())
            hand_ok = detector_.is_left_fist_exclusive(lh, &rh);
        else if (pose_id == 5 && !rh.empty() && !lh.empty())
            hand_ok = detector_.are_both_fists_closed(rh, lh);

        if (pose_ok || hand_ok)
        {
            float total = std::chrono::duration<float>(std::chrono::steady_clock::now() - start).count();
            RCLCPP_INFO(this->get_logger(), "✅ Postura detectada correctamente en %.2f s", total);
            publish_result(true, total);
            return;
        }

        std::this_thread::sleep_for(100ms);
    }
}

void CheckerNode::publish_result(bool result, float duration)
{
    result_pub_->publish(std_msgs::msg::Bool().set__data(result));
    duration_pub_->publish(std_msgs::msg::Float32().set__data(duration));
    RCLCPP_INFO(this->get_logger(), "📤 Resultado publicado: %s (%.2fs)",
                result ? "✅ Correcto" : "❌ Incorrecto", duration);
}

}  // namespace posture_game

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<posture_game::CheckerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
